physics/clm_lake.f90

```diff
@@ -0,0 +1,5737 @@
+
+         ! Error reporting:
+         errflg, errmsg)
+
+      !==============================================================================
+      ! This subroutine was first edited by Hongping Gu and Jiming Jin for coupling
+      ! 07/20/2010
+      ! Long after, in June 2022, Sam Trahan updated it for CCPP
+      !==============================================================================
+
+    IMPLICIT NONE
+    
+    !
+    ! Model time and metadata:
+    !
+    INTEGER , INTENT (IN) :: im,km,me,master
+    INTEGER, INTENT(IN) :: IDATE(4), kdt
+    REAL, INTENT(IN) :: fhour
+
+    !
+    ! Configuration and initialization:
+    !
+    INTEGER, INTENT(IN) :: iopt_lake, iopt_lake_clm
+    REAL(KIND_PHYS), INTENT(IN)  :: min_lakeice, lakedepth_default, dtp
+    LOGICAL, INTENT(IN) :: use_lakedepth
+    INTEGER, DIMENSION(:), INTENT(IN) :: use_lake_model
+    REAL(KIND_PHYS), INTENT(INOUT) :: clm_lake_initialized(:)
+    LOGICAL, INTENT(IN) :: frac_grid, frac_ice
+
+    !
+    ! Atmospheric model state inputs:
+    !
+    REAL(KIND_PHYS), DIMENSION(:), INTENT(IN):: &
+         tg3, pgr, zlvl, qvcurr, xlat_d, xlon_d, ch, cm, &
+         dlwsfci, dswsfci, emiss, oro_lakedepth, wind, rho0, tsfc, &
+         rainncprv, raincprv
+    REAL(KIND_PHYS), DIMENSION(:,:), INTENT(in) :: gu0, gv0, prsi, gt0, phii
+    LOGICAL, DIMENSION(:), INTENT(IN) :: flag_iter
+    INTEGER, DIMENSION(:), INTENT(IN) :: ISLTYP
+
+    !
+    ! Feedback to atmosphere:
+    !
+    REAL(KIND_PHYS), DIMENSION(:), INTENT(INOUT) :: &
+         evap_wat,     evap_ice,   hflx_wat,    hflx_ice,  gflx_wat, gflx_ice,    &
+         ep1d_water,   ep1d_ice,   tsurf_water, tsurf_ice, tsfc_wat, tisfc,       &
+         weasdi,       snodi,      hice,        qss_water, qss_ice,               &
+         cmm_water,    cmm_ice,    chh_water,   chh_ice,                          &
+         uustar_water, uustar_ice, lake_t_snow, albedo,    zorlw,                 &
+         zorli,        lake_t2m,   lake_q2m,    weasd,     snowd,    fice
+    LOGICAL, INTENT(INOUT) :: icy(:)
+
+    !
+    ! Lake model internal state stored by caller:
+    !
+    INTEGER, DIMENSION( : ), INTENT(INOUT)    :: salty
+
+    real(kind_phys),           dimension(: )                ,intent(inout)  :: savedtke12d,    &
+                                                                               snowdp2d,       &    
+                                                                               h2osno2d,       &    
+                                                                               snl2d,          &    
+                                                                               t_grnd2d
+    
+    real(kind_phys),    dimension( :,: )           ,INTENT(inout)  :: t_lake3d,       &    
+                                                                                  lake_icefrac3d
+    real(kind_phys),    dimension( :,-nlevsnow+1: )  ,INTENT(inout)  :: t_soisno3d,     &    
```

Comment:
You are mistaken. This is a differed shape array since there is no upper bound. The variable has to have a lower bound for the code to work. Specifying a lower bound does not make it cease to be a deferred shape array.

Reply:
@dustinswales I have left a comment in the code to educate reviewers on the CCPP design flaw that forces us to use deferred shape arrays, and how a deferred shape array is declared. Hopefully I won't get any more requests to change my deferred shape arrays to something else.

```f90
! Quick education on CCPP and deferred shape arrays.                                                                                                                                                                                         

! CCPP requires deferred shape arrays as a workaround for its design                                                                                                                                                                         
! flaw: it needs an argument that can receive either a null pointer,                                                                                                                                                                         
! or an automatic storage array (which is not guaranteed to exist in                                                                                                                                                                         
! memory at all. Such a thing doesn't exist in Fortran, so the design                                                                                                                                                                        
! of CCPP assumes a compiler will accept either as an argument to a                                                                                                                                                                          
! deferred shape array.                                                                                                                                                                                                                      

! Apparently there is a misunderstanding among developers of how a                                                                                                                                                                           
! deferred shape array is declared. If the array dimensions do not                                                                                                                                                                           
! have an UPPER bound, then it is deferred shape. A LOWER bound is                                                                                                                                                                           
! acceptable; it does not cease to be a deferred shape array.                                                                                                                                                                                

! That is why these seven arrays fit the CCPP design.                                                                                                                                                                                        
```

Reply:
IMO, this is inappropriate to have in a code base. If additional documentation is needed, I think it belongs in section 2.4 of https://ccpp-techdoc.readthedocs.io/en/latest/CompliantPhysicsParams.html that already mentions that lower bounds are fine.

Reply:
@SamuelTrahanNOAA Point taken, but please remove this.

Reply:
It's gone.